known_drivers/mk-nvidia-drivers.py

```diff
@@ -0,0 +1,69 @@
+import requests
+from bs4 import BeautifulSoup
+import re
+import subprocess
+from subprocess import CalledProcessError
+import os
+import shutil
+
+CLEANUP=True
+
+headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'}
+
+nv = requests.get("https://download.nvidia.com/XFree86/Linux-x86_64/", headers=headers)
+
+soup = BeautifulSoup(nv.content.decode(), 'html.parser')
+dirs = [e.text for e in soup.select(".dir")]
+vers = [re.sub("/$", "", t) for t in dirs if not re.search("^..$|-", t)]
+
+def process_output(o):
+    lines = o.decode().split("\n")
+    pth_line = lines[0]
+    store_pth = re.search("'(.*)'", pth_line).group(1)
+    store_hash = lines[1]
+
+    return (store_pth, store_hash)
+
+if os.path.exists("driver-versions.nix"):
+    print("driver-versions.nix already exists, parsing known versions to skip re-downloads.")
+    print("copying old versions to driver-versions.bak.nix.")
+    shutil.copyfile("driver-versions.nix", "driver-versions.bak.nix")
+    known_versions = []
+    old_version_lines = []
+    with open("driver-versions.nix") as f:
+          for line in f:
+            if line.startswith("{"):
+                version = re.search('version = "([^"]*)"', line).group(1)
+                known_versions.append(version)
+                old_version_lines.append(line)
+
+with open("driver-versions.nix", "wt") as f:
+    print("[", file=f)
+    for line in old_version_lines:
+        print(line.strip(), file=f)
+
+    for v in vers:
+        if v not in known_versions: 
+            try: 
+                url = f"https://download.nvidia.com/XFree86/Linux-x86_64/{v}/NVIDIA-Linux-x86_64-{v}.run"
+                print(f"Trying download of {url}")
+                out = subprocess.check_output(["nix-prefetch-url", url], stderr = subprocess.STDOUT)
+                store_pth, store_hash = process_output(out)
+                if CLEANUP:
+                    print(f"Attempting cleanup of {store_pth}")
+                    try:
+                        subprocess.check_output(["nix-store", "--delete", store_pth])
+                        print("Cleanup success.")
```

Comment:
Why not just run nix-collect-garbage?

Reply:
in my own nix use I'm not very careful with my gcroots and would prefer the gc not be run unless I do it explicitly. Do you see advantages to using that instead?

Reply:
It just feels unnecessary if there are easier ways.

Reply:
do you mean running nix-collect-garbage after downloading each runfile? If so I don't see if as much of an improvement and may inadvertently gc something unintended. 

If you mean wait until the end of the script and then call gc, if so that's probably a bit faster, the store deletion adds some overhead to the download loop, but still runs the risk of gc'ing something unintended. Further the runfiles are now quite large and on computers with limited space waiting to the end to gc might not be best, this was why I wrote this in in the first place.

If you mean leave it up to the user to GC or not after running this script, that sounds good, suffers the space issue noted above, but in most contexts, especially with incremental downloading is probably fine.

Reply:
> do you mean running nix-collect-garbage after downloading each runfile? If so I don't see if as much of an improvement and may inadvertently gc something unintended.

Just once in the end if desired and I would be willing to accept that.

Reply:
I'm happy with this approach; it seems more fine-grained and more performant than calling nix-collect-garbage, and you could always disabled cleanup and call the gc yourself if preferred.